File: APP03022-mrch.backend.somx.catalogos-api/src/main/java/com/sodimac/catman/api/controller/SupplierController.java
package com.sodimac.catman.api.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.sodimac.catman.api.exception.ExceptionWrapper;
import com.sodimac.catman.api.model.dto.PaymentConditionDto;
import com.sodimac.catman.api.model.dto.SupplierCreateDto;
import com.sodimac.catman.api.model.dto.SupplierDto;
import com.sodimac.catman.api.model.dto.SupplierFilterDto;
import com.sodimac.catman.api.model.dto.SupplierTypeDto;
import com.sodimac.catman.api.model.dto.SupplierUpdateDto;
import com.sodimac.catman.api.service.SupplierService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;

@RestController
@RequestMapping("/suppliers")
@Tag(name = "Proveedores", description = "API para gestion de proveedores")
public class SupplierController {

    private final SupplierService supplierService;

    public SupplierController(SupplierService supplierService) {
        this.supplierService = supplierService;
    }

    @Operation(
        operationId = "getAllSuppliers",
        summary = "Obtiene todos los proveedores",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de proveedores",
                content = @Content(schema = @Schema(implementation = SupplierDto[].class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping
    public ResponseEntity<List<SupplierDto>> getAllSuppliers(
            @Parameter(description = "Estado: 1=Activo, 0=Inactivo. Por defecto: solo activos")
            @RequestParam(required = false) Integer status) {
        List<SupplierDto> suppliers = status != null
                ? supplierService.findByStatus(status)
                : supplierService.findAll();
        return ResponseEntity.ok(suppliers);
    }

    @Operation(
        operationId = "getSupplierById",
        summary = "Obtiene un proveedor por ID",
        responses = {
            @ApiResponse(responseCode = "200", description = "Proveedor encontrado",
                content = @Content(schema = @Schema(implementation = SupplierDto.class))),
            @ApiResponse(responseCode = "404", description = "Proveedor no encontrado",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/{id}")
    public ResponseEntity<SupplierDto> getSupplierById(
            @Parameter(description = "ID del proveedor", required = true)
            @PathVariable Integer id) {
        return supplierService.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @Operation(
        operationId = "getSupplierByNumber",
        summary = "Obtiene un proveedor por numero",
        responses = {
            @ApiResponse(responseCode = "200", description = "Proveedor encontrado",
                content = @Content(schema = @Schema(implementation = SupplierDto.class))),
            @ApiResponse(responseCode = "404", description = "Proveedor no encontrado",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/number/{supplierNumber}")
    public ResponseEntity<SupplierDto> getSupplierByNumber(
            @Parameter(description = "Numero de proveedor", required = true)
            @PathVariable String supplierNumber) {
        return supplierService.findBySupplierNumber(supplierNumber)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @Operation(
        operationId = "getSupplierByRfc",
        summary = "Obtiene un proveedor por RFC",
        responses = {
            @ApiResponse(responseCode = "200", description = "Proveedor encontrado",
                content = @Content(schema = @Schema(implementation = SupplierDto.class))),
            @ApiResponse(responseCode = "404", description = "Proveedor no encontrado",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/rfc/{rfc}")
    public ResponseEntity<SupplierDto> getSupplierByRfc(
            @Parameter(description = "RFC del proveedor", required = true)
            @PathVariable String rfc) {
        return supplierService.findByRfc(rfc)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @Operation(
        operationId = "createSupplier",
        summary = "Crea un nuevo proveedor",
        responses = {
            @ApiResponse(responseCode = "201", description = "Proveedor creado",
                content = @Content(schema = @Schema(implementation = SupplierDto.class))),
            @ApiResponse(responseCode = "400", description = "Datos invalidos",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @PostMapping
    public ResponseEntity<SupplierDto> createSupplier(
            @Valid @RequestBody SupplierCreateDto dto,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId) {
        SupplierDto created = supplierService.create(dto, userId);
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    @Operation(
        operationId = "updateSupplier",
        summary = "Actualiza un proveedor existente",
        responses = {
            @ApiResponse(responseCode = "200", description = "Proveedor actualizado",
                content = @Content(schema = @Schema(implementation = SupplierDto.class))),
            @ApiResponse(responseCode = "404", description = "Proveedor no encontrado",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Datos invalidos",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @PutMapping("/{id}")
    public ResponseEntity<SupplierDto> updateSupplier(
            @Parameter(description = "ID del proveedor", required = true)
            @PathVariable Integer id,
            @Valid @RequestBody SupplierUpdateDto dto,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId) {
        return supplierService.update(id, dto, userId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @Operation(
        operationId = "deleteSupplier",
        summary = "Elimina (desactiva) un proveedor",
        responses = {
            @ApiResponse(responseCode = "204", description = "Proveedor eliminado"),
            @ApiResponse(responseCode = "404", description = "Proveedor no encontrado",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteSupplier(
            @Parameter(description = "ID del proveedor", required = true)
            @PathVariable Integer id,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId) {
        boolean deleted = supplierService.delete(id, userId);
        return deleted
                ? ResponseEntity.noContent().build()
                : ResponseEntity.notFound().build();
    }

    @Operation(
        operationId = "getAllSupplierTypes",
        summary = "Obtiene todos los tipos de proveedor",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de tipos de proveedor",
                content = @Content(schema = @Schema(implementation = SupplierTypeDto[].class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/types")
    public ResponseEntity<List<SupplierTypeDto>> getAllSupplierTypes() {
        return ResponseEntity.ok(supplierService.findAllSupplierTypes());
    }

    @Operation(
        operationId = "getAllPaymentConditions",
        summary = "Obtiene todas las condiciones de pago",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de condiciones de pago",
                content = @Content(schema = @Schema(implementation = PaymentConditionDto[].class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/payment-conditions")
    public ResponseEntity<List<PaymentConditionDto>> getAllPaymentConditions() {
        return ResponseEntity.ok(supplierService.findAllPaymentConditions());
    }

    @Operation(
        operationId = "filterSuppliers",
        summary = "STM-831: Filtra proveedores por tipo y estado de bloqueo",
        description = "Permite filtrar proveedores por tipo (0=todos, 1=mercancia, 2=transporte, 3=indirectos, 4=servicios) " +
                     "y por estado de bloqueo (0=solo bloqueados, 1=solo activos, 2=todos)",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de proveedores filtrados",
                content = @Content(schema = @Schema(implementation = SupplierFilterDto[].class))),
            @ApiResponse(responseCode = "400", description = "Tipo de proveedor invalido",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class))),
            @ApiResponse(responseCode = "500", description = "Error interno",
                content = @Content(schema = @Schema(implementation = ExceptionWrapper.class)))
        })
    @GetMapping("/filter")
    public ResponseEntity<List<SupplierFilterDto>> filterSuppliers(
            @Parameter(description = "Tipo de proveedor: 0=Todos, 1=Mercancia, 2=Transporte, 3=Indirectos, 4=Servicios", required = true)
            @RequestParam Integer tipoProveedor,
            @Parameter(description = "Estado de bloqueo: 0=Solo bloqueados, 1=Solo activos, 2=Todos", required = true)
            @RequestParam Integer estatusBloqueo) {
        List<SupplierFilterDto> suppliers = supplierService.findByTypeAndBlockStatus(tipoProveedor, estatusBloqueo);
        return ResponseEntity.ok(suppliers);
    }
}

File: APP03022-mrch.backend.somx.util-api/src/controllers/supplier.controller.ts
import type { Request, Response, NextFunction } from 'express';
import * as supplierService from '@/services/supplier.service.js';
import { SupplierCreateSchema, SupplierUpdateSchema } from '@/dto/supplier.dto.js';
import { resolveUserKey } from '@/middlewares/permission.middleware.js';
import * as securityService from '@/services/security.service.js';
import type { SupplierSecurityFilter } from '@/repositories/supplier.repo.js';

function userId(req: Request): string {
    const h = req.header('X-User-Id');
    return h && h.trim() !== '' ? h : 'system';
}

async function resolveSupplierSecurity(req: Request): Promise<SupplierSecurityFilter | undefined> {
    const userKey = resolveUserKey(req);
    if (!userKey) return undefined;
    try {
        const data = await securityService.getUserAttributesByKey(userKey);
        const attrs = data?.attributes ?? [];
        const vendors = attrs
            .filter((a) => a.typeKey === 'ATR001')
            .map((a) => a.valueKey)
            .filter((v): v is string => typeof v === 'string' && v.trim() !== '');
        const typeIds = attrs
            .filter((a) => a.typeKey === 'ATR002')
            .map((a) => Number(String(a.valueKey ?? '').replace(/\D/g, '')))
            .filter((n) => !Number.isNaN(n) && n > 0);
        return {
            vendors: vendors.length > 0 ? vendors : null,
            typeIds: typeIds.length > 0 ? typeIds : null,
        };
    } catch {
        return undefined;
    }
}

export async function getAllSuppliers(req: Request, res: Response, next: NextFunction) {
    try {
        const statusRaw = req.query.status;
        const security = await resolveSupplierSecurity(req);
        const suppliers =
            statusRaw !== undefined && statusRaw !== null && statusRaw !== ''
                ? await supplierService.findByStatus(Number(statusRaw), security)
                : await supplierService.findAll(security);
        res.json(suppliers);
    } catch (err) {
        next(err);
    }
}

export async function getSupplierById(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const result = await supplierService.findById(id);
        if (!result) return res.status(404).json({ error: 'Proveedor no encontrado' });
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function getSupplierByNumber(req: Request, res: Response, next: NextFunction) {
    try {
        const result = await supplierService.findBySupplierNumber(String(req.params.supplierNumber));
        if (!result) return res.status(404).json({ error: 'Proveedor no encontrado' });
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function getSupplierTypeBlocked(req: Request, res: Response, next: NextFunction) {
    try {
        const supplierNumber = String(req.params.supplierNumber);
        const typeBlocked = await supplierService.isSupplierTypeBlocked(supplierNumber);
        res.json({ supplierNumber, typeBlocked });
    } catch (err) {
        next(err);
    }
}

export async function getSupplierByRfc(req: Request, res: Response, next: NextFunction) {
    try {
        const result = await supplierService.findByRfc(String(req.params.rfc));
        if (!result) return res.status(404).json({ error: 'Proveedor no encontrado' });
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function createSupplier(req: Request, res: Response, next: NextFunction) {
    try {
        const dto = SupplierCreateSchema.parse(req.body);
        const created = await supplierService.create(dto, userId(req));
        res.status(201).json(created);
    } catch (err) {
        next(err);
    }
}

export async function updateSupplier(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const dto = SupplierUpdateSchema.parse(req.body);
        const updated = await supplierService.update(id, dto, userId(req));
        if (!updated) return res.status(404).json({ error: 'Proveedor no encontrado' });
        res.json(updated);
    } catch (err) {
        next(err);
    }
}

export async function deleteSupplier(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const ok = await supplierService.deleteSupplier(id, userId(req));
        if (!ok) return res.status(404).json({ error: 'Proveedor no encontrado' });
        res.status(204).send();
    } catch (err) {
        next(err);
    }
}

export async function getAllSupplierTypes(_req: Request, res: Response, next: NextFunction) {
    try {
        const result = await supplierService.findAllSupplierTypes();
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function getAllPaymentConditions(_req: Request, res: Response, next: NextFunction) {
    try {
        const result = await supplierService.findAllPaymentConditions();
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function filterSuppliers(req: Request, res: Response, next: NextFunction) {
    try {
        const tipoProveedor = Number(req.query.tipoProveedor);
        const estatusBloqueo = Number(req.query.estatusBloqueo);
        if (Number.isNaN(tipoProveedor) || Number.isNaN(estatusBloqueo)) {
            return res.status(400).json({ error: 'Parámetros inválidos' });
        }
        const security = await resolveSupplierSecurity(req);
        const result = await supplierService.findByTypeAndBlockStatus(tipoProveedor, estatusBloqueo, security);
        res.json(result);
    } catch (err) {
        next(err);
    }
}

